fedimint-core/src/db/mem_impl.rs

```diff
@@ -176,21 +180,25 @@ impl<'a> IDatabaseTransactionOps for MemTransaction<'a> {
 #[apply(async_trait_maybe_send!)]
 impl<'a> IRawDatabaseTransaction for MemTransaction<'a> {
     async fn commit_tx(self) -> Result<()> {
+        let mut data = self.db.data.lock().await;
```

Comment:
If you move the lock all the way to the top, what's the point of comparing old values. Nothing could have changed them. It's the read-write conflicts that might break snapshot level isolation.

Reply:
imagine 
tx a: delete(x)
tx b: delete(x)

if we don't compare old values, then commit both transactions will succeed.